File: src/main/java/io/scriptor/loader/ITree.java
package io.scriptor.loader;

import org.jetbrains.annotations.NotNull;

import java.util.*;

public interface ITree<T> extends Iterable<T> {

    Iterator<ITree<T>> children();

    T value();

    boolean hasValue();

    @Override
    default @NotNull Iterator<T> iterator() {
        return new TreeIterator<>(this);
    }

    final class TreeIterator<T> implements Iterator<T> {

        private final Deque<Iterator<ITree<T>>> stack = new ArrayDeque<>();
        private T next;

        public TreeIterator(final @NotNull ITree<T> root) {
            this.stack.push(Collections.singletonList(root).iterator());
            advance();
        }

        @Override
        public boolean hasNext() {
            return next != null;
        }

        @Override
        public T next() {
            if (!hasNext()) {
                throw new NoSuchElementException();
            }
            final var result = next;
            advance();
            return result;
        }

        private void advance() {
            next = null;

            while (!stack.isEmpty()) {
                final var top = stack.peek();

                if (top.hasNext()) {
                    final var current = top.next();
                    stack.push(current.children());

                    if (current.hasValue()) {
                        next = current.value();
                        return;
                    }
                } else {
                    stack.pop();
                }
            }
        }
    }
}

File: src/main/java/io/scriptor/loader/ITree.kt
package io.scriptor.loader

import java.util.*

interface ITree<T> : Iterable<T> {

    val children: MutableIterator<ITree<T>>

    val value: T?

    fun hasValue(): Boolean

    override fun iterator(): MutableIterator<T> {
        return TreeIterator(this)
    }

    class TreeIterator<T>(root: ITree<T>) : MutableIterator<T> {
        private val stack: Deque<MutableIterator<ITree<T>>> = ArrayDeque()
        private var next: T? = null

        init {
            this.stack.push(mutableListOf(root).iterator())
            advance()
        }

        override fun hasNext(): Boolean {
            return next != null
        }

        override fun next(): T {
            val result = next ?: throw NoSuchElementException()
            advance()
            return result
        }

        private fun advance() {
            next = null

            while (!stack.isEmpty()) {
                val top = stack.peek()

                if (top.hasNext()) {
                    val current = top.next()
                    stack.push(current.children)

                    if (current.hasValue()) {
                        next = current.value
                        return
                    }
                } else {
                    stack.pop()
                }
            }
        }

        override fun remove() {
            throw UnsupportedOperationException()
        }
    }
}
